design/Merkle.md

```diff
@@ -0,0 +1,226 @@
+  functions like SHA256), we can just concatenate the two leaves of the node and apply the
+  hash: `compress(x,y) := H(x|y)`
+- in case of hash functions based on the sponge construction (like Poseidon or Keccak/SHA3), 
+  we can just fill the "capacity part" of the state with a constant (say 0), the "absorbing 
+  part" of the state with the two inputs, apply the permutation, and extract a single `T` 
+
+### Attacks
+
+When implemented without enough care (like the above naive algorithm), there are several 
+possible attacks producing hash collisions or second preimages:
+
+1. The root of particular any layer is the same as the root of the input
+2. The root of `[x_0,x_1,...,x_(2*k)]` (length is `n=2*k+1` is the same as the root of 
+   `[x_0,x_1,...,x_(2*k),dummy]` (length is `n=2*k+2`)
+3. when using bytes as the input, already `deserialize` can have similar collision attacks
+4. The root of a singleton sequence is itself
+
+Traditional (linear) hash functions usually solve the analogous problems by clever padding.
+
+### Domain separation
+
+It's a good practice in general to ensure that different constructions using the same 
+underlying hash functions will never produce the same output. This is called "domain separation",
+and it's a bit similar to _multihash_; however instead of adding extra bits of information to a hash
+(and thus increasing its size), we just compress the extra information into the hash itself.
+A simple example would be using `H(dom|H(...))` instead of `H(...)`. The below solutions
+can be interpreted as an application of this idea, where we want to separate the different
+lengths `n`.
+
+### Possible solutions (for the tree attacks)
+
+While the third problem (`deserialize` may be not injective) is similar to the second problem,
+let's deal first with the tree problems, and come back to `deserialize` (see below) later.
+
+**Solution 0b.** Pre-hash each input element. This solves 2) and 4) (if we choose `dummy` to be
+something we don't expect anybody to find a preimage), but does not solve 1); also it
+doubles the computation time.
+
+**Solution 1.** Just prepend the data with the length `n` of the input sequence. Note that any
+cryptographic hash function needs an output size of at least 160 bits (and usually at least 
+256 bits), so we can always embed the length (surely less than `2^64`) into `T`. This solves
+both problems 1) and 2) (the height of the tree is a deterministic function of the length),
+and 4) too.
+However, a typical application of a Merkle tree is the case where the length of the input
+`n=2^d` is a power of two; in this case it looks a little bit "inelegant" to increase the size
+to `n=2^d+1`, though the overhead with above even-odd construction is only `log2(n)`.
+An advantage is that you can _prove_ the size of the input with a standard Merkle inclusion proof.
+Alternative version: append instead of prepend; then the indexing of the leaves does not change.
```

Comment:
Appending has another advantage; it makes it easier to deal with scenario's where the length isn't known beforehand, such as when streaming the data.